packages/protocol/contracts/bridge/erc721/libs/LibERC721.sol

```diff
@@ -0,0 +1,227 @@
+        string canonicalTokenSymbol,
+        string canonicalTokenName
+    );
+
+    event ERC721Sent(
+        bytes32 indexed msgHash,
+        address indexed from,
+        address indexed to,
+        uint256 destChainId,
+        address token,
+        uint256 tokenId
+    );
+
+    event ERC721Released(
+        bytes32 indexed msgHash,
+        address indexed from,
+        address token,
+        uint256 tokenId
+    );
+
+    event ERC721Received(
+        bytes32 indexed msgHash,
+        address indexed from,
+        address indexed to,
+        uint256 srcChainId,
+        address token,
+        uint256 tokenId
+    );
+
+    error NFTVAULT_CANONICAL_TOKEN_NOT_FOUND();
+    error NFTVAULT_INVALID_OWNER();
+    error NFTVAULT_INVALID_SENDER();
+    error NFTVAULT_INVALID_SRC_CHAIN_ID();
+    error NFTVAULT_INVALID_TOKEN();
+    error NFTVAULT_MESSAGE_NOT_FAILED();
+
+    function sendToken(
+        address owner,
+        address to,
+        uint256 tokenId,
+        address token,
+        string memory tokenUri,
+        bool isBridgedToken,
+        NFTVaultParent.CanonicalNFT memory bridgedToCanonical,
+        bytes4 selector
+    ) public returns (bytes memory) {
+        NFTVaultParent.CanonicalNFT memory canonicalToken;
+
+        // is a bridged token, meaning, it does not live on this chain
+        if (isBridgedToken) {
+            if (BridgedERC721(token).ownerOf(tokenId) != msg.sender)
+                revert NFTVAULT_INVALID_OWNER();
+
+            BridgedERC721(token).bridgeBurnFrom(msg.sender, tokenId);
+            canonicalToken = bridgedToCanonical;
+            if (canonicalToken.tokenAddr == address(0))
+                revert NFTVAULT_CANONICAL_TOKEN_NOT_FOUND();
+        } else {
+            // is a canonical token, meaning, it lives on this chain
+            ERC721Upgradeable t = ERC721Upgradeable(token);
+            if (t.ownerOf(tokenId) != msg.sender)
+                revert NFTVAULT_INVALID_OWNER();
+
+            canonicalToken = NFTVaultParent.CanonicalNFT({
+                srcChainId: block.chainid,
+                tokenAddr: token,
+                symbol: t.symbol(),
+                name: t.name(),
+                uri: tokenUri
+            });
+            t.transferFrom(msg.sender, address(this), tokenId);
+        }
+
+        return
+            abi.encodeWithSelector(
+                selector,
+                canonicalToken,
+                owner,
+                to,
+                tokenId
+            );
+    }
+
+    function receiveToken(
+        AddressResolver resolver,
+        address addressManager,
+        NFTVaultParent.CanonicalNFT memory canonicalToken,
+        address from,
+        address to,
+        uint256 tokenId,
+        address canonicalToBridged
+    ) public returns (bool bridged, address token) {
+        IBridge.Context memory ctx = IBridge(msg.sender).context();
+        if (ctx.sender != resolver.resolve(ctx.srcChainId, "erc721_vault", false))
```

Comment:
Line 125 to 127 are common logics to all vaults, we should move it to the base class (parent). Using the lib won't allow us to do it, I suggest that we remove the Libs.